### PR TITLE
 Add padding='same' mode to conv{1,2,3}d, nn.Conv{1,2,3}d

### PR DESCRIPTION
**Pull request by peterbell10**
First part of #3867 (Pooling operators still to do)

This adds a `padding='same'` mode to the interface of `conv{n}d`and `nn.Conv{n}d`. This should match the behaviour of `tensorflow`. I couldn't find it explicitly documented but through experimentation I found `tensorflow` returns the shape `ceil(len/stride)` and always adds any extra asymmetric padding onto the right side of the input.

Since the `native_functions.yaml` schema doesn't seem to support strings or enums, I've moved the function interface into python and it now dispatches between the numerically padded `conv{n}d` and the `_conv{n}d_same` variant. Underscores because I couldn't see any way to avoid exporting a function into the `torch` namespace. 

A note on asymmetric padding. The total padding required can be odd if both the kernel-length is even  and the dilation is odd. mkldnn has native support for asymmetric padding, so there is no overhead there, but for other backends I resort to padding the input tensor by 1 on the right hand side to make the remaining padding symmetrical. In these cases, I use `TORCH_WARN_ONCE` to notify the user of the performance implications.

### CODE REVIEW

File: torch/nn/functional.py [hunk 1]
```diff
@@ -11,121 +11,266 @@
...
+        if padding == 'same':
+            return _VF._conv1d_same(input, weight, bias, stride, dilation, groups)
+        if padding == 'valid':
+            return _VF.conv1d(input, weight, bias, stride, padding=0,
+                              dilation=dilation, groups=groups)
+        else:
+            raise ValueError('Invalid padding mode {!r}'.format(padding))
+    else:
+        # FIXME: JIT doesn't understand this
+        assert not isinstance(padding, str)
```
Comment (peterbell10) [inline torch/nn/functional.py:97]: The JIT won't doesn't support `Union` so instead of `Union[str, BroadcastingList1[int]]` I have to use two `_overload`s with different types for `padding`. This doesn't compile though because, while the JIT understands the `if isinstance` branch must contain a string, it doesn't understand that the `else` branch must _not_ contain a string. So, JIT fails because `_VF.conv1d` doesn't support a string `padding`.

This could be fixed either in the JIT, or by updating the `native_functions.yaml` schema to support string types. That is, unless I'm missing something and there's already a way to support JIT here.

@ezyang any thoughts on this?

Comment (ezyang) [reply torch/nn/functional.py:97]: I don't know. @suo any ideas?

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -802,12 +802,32 @@
...
 
 - func: conv2d(Tensor input, Tensor weight, Tensor? bias=None, int[2] stride=1, int[2] padding=0, int[2] dilation=1, int groups=1) -> Tensor
 
 - func: conv3d(Tensor input, Tensor weight, Tensor? bias=None, int[3] stride=1, int[3] padding=0, int[3] dilation=1, int groups=1) -> Tensor
 
+- func: conv1d.padding_modeA(Tensor input, Tensor weight, Tensor bias, int[1] stride, str padding, int[1] dilation=1, int groups=1) -> Tensor
+
+- func: conv1d.padding_modeB(Tensor input, Tensor weight, Tensor bias, *, str padding, int[1] stride=1, int[1] dilation=1, int groups=1) -> Tensor
+
+- func: conv1d.padding_modeC(Tensor input, Tensor weight, *, str padding, Tensor? bias=None, int[1] stride=1, int[1] dilation=1, int groups=1) -> Tensor
```
Comment (peterbell10) [inline aten/src/ATen/native/native_functions.yaml:817]: Okay, turns out I was mistaken and `native_functions.yaml` does indeed support `str`. It was actually failing to compile because it won't accept `str padding` as not having a default while the earlier arguments `bias` and `stride` do.

However, if `padding` is given a default value it can be ambiguous with the `int[1]` overload and C++ code fails to compile. At the same time, `bias` and `stride` need to have defaults in python for hen you call it like `conv1d(input, weight, padding='same')`. I found this ugly workaround of defining these two extra overloads where padding is keyword-only and to the left of all the defaulted arguments.

File: aten/src/ATen/native/mkldnn/Conv.cpp [hunk 3]
```diff
@@ -248,7 +242,8 @@ std::tuple<at::Tensor,at::Tensor,at::Tensor> mkldnn_convolution_backward(
     IntArrayRef padding_l, IntArrayRef padding_r, IntArrayRef stride,
     IntArrayRef dilation, int64_t groups, std::array<bool, 3> output_mask)
 {
-  Tensor grad_output = grad_output_t.contiguous();
+  const Tensor& grad_output = grad_output_t.is_mkldnn() ?
+      grad_output_t : grad_output_t.contiguous();
```
Comment (peterbell10) [inline aten/src/ATen/native/mkldnn/Conv.cpp:246]: `Tensor.contiguous()` here doesn't work for `mkldnn` tensors. There are also other places where the `grad`s are all returned as CPU tensors and autograd will raise an error about expecting `layout=Mkldnn`.

It seems the mkldnn backend is not actually tested properly with mkldnn tensors as inputs. Is that not the expected use case?

File: aten/src/ATen/native/native_functions.yaml (hunk 2 above)
Comment (ezyang) [reply aten/src/ATen/native/native_functions.yaml:838]: ooooooof

**Comment by ezyang** [comment]
The JIT schema changes look... pretty gnarly. We may need to figure out if there is some core changes we should make first to support this addition.

**Comment by eellison** [comment]
what do ppl think of the approach in https://github.com/pytorch/pytorch/pull/22484/files ?

**Comment by peterbell10** [comment]
Oh, I somehow missed that PR reading through the issue. The obvious problem is that by calculating the padding in python, there is no support in the C++ `torch` api.

**Comment by peterbell10** [comment]
>We're going to talk about this at the composability meeting today

@ezyang what was the outcome of that discussion?

**Comment by ezyang** [comment]
@bhosmer signed up to scribe here

**Comment by bhosmer** [comment]
> @bhosmer signed up to scribe here

Confirming - I'll put up a summary of our current thinking in the next day or so.

**Comment by bhosmer** [comment]
HI @peterbell10, many apologies for the delayed recap. Here are the main points we arrived at in that discussion, along with some subsequent thought:
* it'd be nice to not have to have Python wrappers simply to avoid schema defaulting issues: I think the things you're running into (ambiguous overloads, string defaulting) aren't fundamental and should be fixable (tracking issues: #43944, #43945). Solving them should let you specify intuitive overloads of conv{n}d in native_functions.yaml.
* given a choice between enums and strings for representing the padding modes themselves, we wound up thinking strings were the better choice given their lighter weight, as long as the implementations are careful about interning, avoiding repeated comparisons etc.
* a disadvantage to downcoding the mode into actual padding values prior to calling the traced op, as the current PR does (via the `convolution_same` calls in the bodies of `conv{n}d`) is that baking the calculated values into the trace make it no longer size invariant. As for the logistics of plumbing the mode down to the traced op, we had some ideas around adding an overload of `_convolution` which took the padding mode, and then having both the current and new overloads call a (untraced) `__convolution` helper that took the calculated padding values. But this needs a bit more thinking.

I think our next steps are:
* fixing the defaulting and overloading issues you ran into in `native_functions.yaml`
* figuring out how best to wrangle the plumbing of padding mode down to a canonical traced op, rather than calculating 

Also cc @eellison - I don't think we gave your misgivings a full airing in our discussion, can you elaborate a little on the downside of adding the padding-mode overloads for `conv{n}d`?

**Comment by peterbell10** [comment]
>As for the logistics of plumbing the mode down to the traced op, we had some ideas around adding an overload of `_convolution` which took the padding mode, and then having both the current and new overloads call a (untraced) `__convolution` helper that took the calculated padding values. But this needs a bit more thinking.

I think I've found a nice solution in the latest commit. I noticed that the tracer stops at `at::_convolution` and doesn't expand it down to the actual implementation like `at::cudnn_convolution`. So, in a similar way I've added the `at::_convolution_mode` function which takes in a mode string and internally makes the padding values concrete and just calls `at::convolution`. When I trace a simple function call, the graph only has `aten::_convolution_mode` and not any padding values.

If I understand correctly, the tracer will stop at the first function call that isn't in the [`DONT_RECORD_TRACE` list](https://github.com/pytorch/pytorch/blob/ac1f471fe20faf96282b4143e4193759f01a7d41/tools/autograd/gen_variable_type.py#L36-L42) and not expand any deeper. Is that right?

**Comment by peterbell10** [comment]
This PR was getting pretty large with changes touching codegen, convolution and mkldnn-specific changes. So, I've split this into a ghstack of PRs, the main one being #45667
